src/mlpack/methods/neighbor_search/spill_search_impl.hpp

```diff
@@ -0,0 +1,251 @@
+    const MetricType metric) :
+    neighborSearch(naive, singleMode, epsilon, metric),
+    tau(tau),
+    rho(rho),
+    leafSize(leafSize)
+{
+  if (tau < 0)
+    throw std::invalid_argument("tau must be non-negative");
+  if (rho < 0 || rho > 1)
+    throw std::invalid_argument("rho must be in the range [0,1]");
+}
+
+// Clean memory.
+template<typename MetricType,
+         typename MatType,
+         template<typename HyperplaneMetricType> class HyperplaneType,
+         template<typename SplitBoundT, typename SplitMatT> class SplitType>
+SpillSearch<MetricType, MatType, HyperplaneType, SplitType>::
+~SpillSearch()
+{
+  /* Nothing to do */
+}
+
+template<typename MetricType,
+         typename MatType,
+         template<typename HyperplaneMetricType> class HyperplaneType,
+         template<typename SplitBoundT, typename SplitMatT> class SplitType>
+void SpillSearch<MetricType, MatType, HyperplaneType, SplitType>::
+Train(const MatType& referenceSet)
+{
+  if (Naive())
+    neighborSearch.Train(referenceSet);
+  else
+  {
+    // Build reference tree with proper value for tau.
+    Tree* tree = new Tree(referenceSet, tau, leafSize, rho);
+    neighborSearch.Train(tree);
+    // Give the model ownership of the tree.
+    neighborSearch.treeOwner = true;
+  }
+}
+
+template<typename MetricType,
+         typename MatType,
+         template<typename HyperplaneMetricType> class HyperplaneType,
+         template<typename SplitBoundT, typename SplitMatT> class SplitType>
+void SpillSearch<MetricType, MatType, HyperplaneType, SplitType>::
+Train(MatType&& referenceSetIn)
+{
+  if (Naive())
+    neighborSearch.Train(std::move(referenceSetIn));
+  else
+  {
+    // Build reference tree with proper value for tau.
+    Tree* tree = new Tree(std::move(referenceSetIn), tau, leafSize, rho);
+    neighborSearch.Train(tree);
+    // Give the model ownership of the tree.
+    neighborSearch.treeOwner = true;
+  }
+}
+
+template<typename MetricType,
+         typename MatType,
+         template<typename HyperplaneMetricType> class HyperplaneType,
+         template<typename SplitBoundT, typename SplitMatT> class SplitType>
+void SpillSearch<MetricType, MatType, HyperplaneType, SplitType>::
+Train(Tree* referenceTree)
+{
+  neighborSearch.Train(referenceTree);
+}
+
+template<typename MetricType,
+         typename MatType,
+         template<typename HyperplaneMetricType> class HyperplaneType,
+         template<typename SplitBoundT, typename SplitMatT> class SplitType>
+void SpillSearch<MetricType, MatType, HyperplaneType, SplitType>::
+Search(const MatType& querySet,
+       const size_t k,
+       arma::Mat<size_t>& neighbors,
+       arma::mat& distances)
+{
+  if (Naive() || SingleMode())
+    neighborSearch.Search(querySet, k, neighbors, distances);
+  else
+  {
+    // For Dual Tree Search on SpillTrees, the queryTree must be built with non
+    // overlapping (tau = 0).
+    Tree queryTree(querySet, 0 /* tau */, leafSize, rho);
```

Comment:
When I originally designed the `NeighborSearch` class, my intention was to make it as tree-independent as possible (or, well, that's what happened in the end).  So it doesn't hold any `leafSize` parameter despite the fact that the `BinarySpaceTree` uses it.  The policy has always been, `NeighborSearch` will build your trees with the default parameters, and if you don't want the default parameters, you must pass the trees.  I think that we should continue that policy here, so I don't think it's advisable to have `leafSize`, `rho`, and `tau` stored in the class.  If the user wants to specify those, they can build the tree themselves.  (This means that `NSModel` will need to store those parameters, if they can be specified from the command-line.)

For the traversers, an idea is to add two more template parameters to `NeighborSearch` that specify the desired single-tree and dual-tree traversers (and default to the defaults of the TreeType).

Then we could typedef `SpillTreeKNN = NeighborSearch<EuclideanDistance, arma::mat, SPTree, SPTree::DefeatistSingleTreeTraverser, SPTree::DefeatistDualTreeTraverser>` (or whatever the names and syntax may be).  This, plus the template specialization I suggested for building the query tree for spill tree search, should be sufficient to avoid the need for the `SpillSearch` class at all, and keep the number of files and classes down.  What do you think?  Have I overlooked a detail?